libcudacxx/include/cuda/__floating_point/fp.h

```diff
@@ -0,0 +1,530 @@
+      : __fp{__construct_from(__v)}
+  {}
+#  endif // ^^^ !_CCCL_NO_CONDITIONAL_EXPLICIT ^^^
+
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(_CUDA_VSTD::is_integral_v<_Tp>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr __fp(const _Tp& __v) noexcept
+      : __fp{__construct_from(__v)}
+  {}
+
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(_CUDA_VSTD::is_same_v<__fp_make_config_from_t<_Tp>, __config_type>)
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr __fp(__fp_from_native_t, const _Tp& __v) noexcept
+      : __fp{__fp_from_storage, _CUDA_VSTD::bit_cast<__storage_type>(__v)}
+  {}
+
+  template <class _Tp>
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr __fp(__fp_from_storage_t, const _Tp& __v) noexcept
+      : __storage_{__v}
+  {
+    static_assert(_CUDA_VSTD::is_same_v<_Tp, __storage_type>);
+  }
+
+#  if defined(_CCCL_NO_CONDITIONAL_EXPLICIT)
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__fp_is_floating_point_v<_Tp> _CCCL_AND __fp_cast_is_implicit<_Tp, __fp>())
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr operator _Tp() const noexcept
+  {
+    return __cast_to<_Tp>();
+  }
+
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__fp_is_floating_point_v<_Tp> _CCCL_AND(!__fp_cast_is_implicit<_Tp, __fp>()))
+  _LIBCUDACXX_HIDE_FROM_ABI explicit constexpr operator _Tp() const noexcept
+  {
+    return __cast_to<_Tp>();
+  }
+#  else // ^^^ _CCCL_NO_CONDITIONAL_EXPLICIT ^^^ / vvv !_CCCL_NO_CONDITIONAL_EXPLICIT vvv
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__fp_is_floating_point_v<_Tp>)
+  _LIBCUDACXX_HIDE_FROM_ABI explicit(!__fp_cast_is_implicit<_Tp, __fp>()) constexpr operator _Tp() const noexcept
+  {
+    return __cast_to<_Tp>();
+  }
+#  endif // ^^^ !_CCCL_NO_CONDITIONAL_EXPLICIT ^^^
+
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(_CUDA_VSTD::is_integral_v<_Tp>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr operator _Tp() const noexcept
+  {
+    return __cast_to<_Tp>();
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __has_native_type() noexcept
+  {
+    NV_IF_ELSE_TARGET(NV_IS_HOST, (return __has_host_native_type;), (return __has_device_native_type;))
+  }
+
+  // private:
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr __fp __construct_from(const _Tp& __v) noexcept
+  {
+#  if defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+    static_assert(!_CUDA_VSTD::is_same_v<_Tp, long double>, "long double is not supported");
+#  endif // _LIBCUDACXX_HAS_NO_LONG_DOUBLE
+
+    // todo: improve the implementation
+    NV_IF_ELSE_TARGET(NV_IS_HOST, (return __construct_from_host<_Tp>(__v);), (return __construct_from_device<_Tp>(__v);))
+  }
+#  if !_CCCL_COMPILER(NVRTC)
+  template <class _Tp, class _FpCast = class __fp_cast>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_HOST static constexpr __fp __construct_from_host(const _Tp& __v) noexcept
+  {
+    if constexpr (__has_host_native_type)
+    {
```

Comment:
Ditto strong proponent of inverting condition and keeping the if else tree small. 

But I am fearful that some combination of toolchains will do the stupid thing